tools/nbs_decoder/__init__.py

```diff
@@ -0,0 +1,114 @@
+            nuclear_dir=nuclear_proto_dir,
+            user_dir=user_proto_dir,
+            proto_files=" ".join(proto_files),
+        )
+    )
+    != 0
+):
+    raise RuntimeError("Unable to build the protocol buffers for decoding")
+
+# Open up our message output directory to get our protobuf types
+sys.path.append(protobuf_path)
+
+# Load all of the protobuf files as modules to use
+for dir_name, subdir, files in os.walk(protobuf_path):
+    modules = pkgutil.iter_modules(path=[dir_name])
+    for loader, module_name, ispkg in modules:
+        if module_name.endswith("pb2"):
+
+            # Load our protobuf module
+            module = loader.find_module(module_name).load_module(module_name)
+
+decoders = {}
+
+# Now that we've imported them all get all the subclasses of protobuf message
+for message in google.protobuf.message.Message.__subclasses__():
+
+    # Reverse to little endian
+    pb_hash = bytes(reversed(xxhash.xxh64(message.DESCRIPTOR.full_name, seed=0x4E55436C).digest()))
+    decoders[pb_hash] = (message.DESCRIPTOR.full_name, message)
+
+
+def decode(path):
+
+    # Now open the passed file
+    with gzip.open(path, "rb") if path.endswith("nbz") or path.endswith(".gz") else open(path, "rb") as f:
+
+        # NBS File Format:
+        # 3 Bytes - NUClear radiation symbol header, useful for synchronisation when attaching to an existing stream
+        # 4 Bytes - The remaining packet length i.e. 16 bytes + N payload bytes
+        # 8 Bytes - 64bit timestamp in microseconds. Note: this is not necessarily a unix timestamp
+        # 8 Bytes - 64bit bit hash of the message type
+        # N bytes - The binary packet payload
+
+        state = enum.Enum("State", "initial header_lock_1 header_lock_2 packet eof")
+        sync_state = state.initial
+
+        # While we can read a header
+        while sync_state != state.eof:
+            # If we reach sync state 3 we are ready to read a packet
+            if sync_state == state.packet:
+                sync_state = state.initial
+                # Read our size
+                size = struct.unpack("<I", f.read(4))[0]
+
+                # Read our payload
+                payload = f.read(size)
+
+                # Read our timestamp
+                timestamp = struct.unpack("<Q", payload[:8])[0]
+
+                # Read our hash
+                type_hash = payload[8:16]
+
+                # If we know how to parse this type, parse it
+                if type_hash in decoders:
+                    # Yield a message
+                    try:
+                        packet = (decoders[type_hash][0], timestamp, decoders[type_hash][1].FromString(payload[16:]))
+                        yield packet
+                    except:
+                        pass
+            else:
+                c = f.read(1)
+                # No bytes to read we are EOF
+                if len(c) == 0:
+                    sync_state = state.eof
+                elif c == b"\xE2":
```

Comment:
State machine does not require a header to exist in its entirety

Reply:
Your critique here is invalid as it was never a state machine in the first place.
On an unrelated note, I've made it a state machine now so it actually works :P